.githooks/pre-commit

```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+set -e
+
+# stash any unstaged changes
+git stash save "prePush" -q --keep-index
+
+# unstash the unstashed changes (if any) on exit or interrupt
+function unstash {
+	git stash apply stash^{/prePush} -q || true
```

Comment:
I think apply leaves the stash, but given this stash is kind of out of your control I feel like having it stay around is a good idea in case the apply doesn't got smoothly.

Perhaps there could be something to drop old stashes?